internal/pkg/nats/options.go

```diff
@@ -0,0 +1,138 @@
+	ClientOptions
+}
+
+// ConnectionOptions contains the connection configurations for the MQTT client.
+//
+// NOTE: The connection properties resides in its own struct in order to avoid the property being loaded in via
+//  reflection during the load process.
+type ConnectionOptions struct {
+	BrokerURL string
+}
+
+// ClientOptions contains the client options which are loaded via reflection
+type ClientOptions struct {
+	// Client Identifiers
+	Username string
+	Password string
+	ClientId string
+	// Connection information
+	// TODO: NATS core is QOS 0, Jetstream QOS 1 (or 2 with sync ack + msgid)
+	// Qos                  int
+	RetryOnFailedConnect bool
+	Durable              bool
+	AutoProvision        bool
+	ConnectTimeout       int // Seconds
+	pkg.TlsConfigurationOptions
+}
+
+// CreateClientConfiguration constructs a ClientConfig based on the provided MessageBusConfig.
+func CreateClientConfiguration(messageBusConfig types.MessageBusConfig) (ClientConfig, error) {
+	var brokerUrl string
+	if !messageBusConfig.PublishHost.IsHostInfoEmpty() {
+		brokerUrl = messageBusConfig.PublishHost.GetHostURL()
+	} else if !messageBusConfig.SubscribeHost.IsHostInfoEmpty() {
+		brokerUrl = messageBusConfig.SubscribeHost.GetHostURL()
+	} else {
+		return ClientConfig{}, fmt.Errorf("Specified empty broker info.")
+	}
+
+	_, err := url.Parse(brokerUrl)
+	if err != nil {
+		return ClientConfig{}, pkg.NewBrokerURLErr(fmt.Sprintf("Failed to parse broker: %v", err))
+	}
+
+	clientOptions := CreateClientOptionsWithDefaults()
+	err = pkg.Load(messageBusConfig.Optional, &clientOptions)
+	if err != nil {
+		return ClientConfig{}, err
+	}
+
+	tlsConfig := pkg.TlsConfigurationOptions{}
+	err = pkg.Load(messageBusConfig.Optional, &tlsConfig)
+	if err != nil {
+		return ClientConfig{}, err
+	}
+
+	clientOptions.TlsConfigurationOptions = tlsConfig
+
+	return ClientConfig{
+		BrokerURL:     brokerUrl,
+		ClientOptions: clientOptions,
+	}, nil
+}
+
+const (
+	format     = "format"
+	QueueGroup = "QueueGroup"
+)
+
+func (cc ClientConfig) ConnectOpt() []nats.Option {
+	connectTimeout := time.Second
+
+	if cc.ConnectTimeout != 0 {
+		connectTimeout = time.Duration(cc.ConnectTimeout) * time.Second
+	}
+
+	opts := []nats.Option{
+		nats.Timeout(connectTimeout),
+		nats.RetryOnFailedConnect(cc.RetryOnFailedConnect),
+	}
+
+	if cc.ClientId != "" {
+		opts = append(opts, nats.Name(cc.ClientId))
+	}
+
+	if cc.Username != "" {
+		opts = append(opts, nats.UserInfo(cc.Username, cc.Password))
+	}
+
+	//TODO: tls config
+
+	return opts
+}
+
+// CreateClientOptionsWithDefaults constructs ClientOptions instance with defaults.
+func CreateClientOptionsWithDefaults() ClientOptions {
+	return ClientOptions{
+		Username:                "",
+		Password:                "",
+		ConnectTimeout:          5, // 5 seconds
+		RetryOnFailedConnect:    false,
+		Durable:                 false,
+		AutoProvision:           false, // AutoProvision JetStream streams - should maybe be true?
```

Comment:
Defaults should be in sync with the NATS defaults if not specified.

Reply:
Default is false - problem there is you then can't publish to a topic until something is listening to it (this is an issue with core NATS also since there is no persistence layer also, there is just no way to auto-provision around it like jetstream provides).

Reply:
Interesting, so we will get an error on publish if no subscriber?